src/replica/bulk_load/replica_bulk_loader.h

```diff
@@ -56,23 +58,30 @@ class replica_bulk_loader : replica_base
 
     // replica start or restart download sst files from remote provider
     // \return ERR_BUSY if node has already had enough replica executing downloading
-    // \return download errors by function `download_sst_files`
+    // \return ERR_FILE_OPERATION_FAILED: create local bulk load dir failed
     error_code start_download(const std::string &remote_dir, const std::string &provider_name);
 
-    // download metadata and sst files from remote provider
+    // download metadata file and create sst download tasks
     // metadata and sst files will be downloaded in {_dir}/.bulk_load directory
-    // \return ERR_FILE_OPERATION_FAILED: create local bulk load dir failed
-    // \return download metadata file error, see function `do_download`
-    // \return parse metadata file error, see function `parse_bulk_load_metadata`
-    error_code download_sst_files(const std::string &remote_dir, const std::string &provider_name);
+    void download_files(const std::string &provider_name,
+                        const std::string &remote_dir,
+                        const std::string &local_dir);
+
+    // download sst files from remote provider
+    void download_sst_file(const std::string &remote_dir,
+                           const std::string &local_dir,
+                           const file_meta &f_meta,
+                           dist::block_service::block_filesystem *fs);
 
     // \return ERR_FILE_OPERATION_FAILED: file not exist, get size failed, open file failed
     // \return ERR_CORRUPTION: parse failed
+    // need to acquire write lock while calling it
```

Comment:
why not put the lock inside these functions?

Reply:
Those functions now will all be locked before calling it, the comments are showing those functions are not thread safety.